src/supernova/test.rs

```diff
@@ -276,26 +273,74 @@ fn print_constraints_name_on_error_index<G1, G2, Ca, Cb>(
 const OPCODE_0: usize = 0;
 const OPCODE_1: usize = 1;
 
-#[derive(Clone, Debug)]
-struct TestROM<G: Group> {
-  op0: CubicCircuit<G::Scalar>,
-  op1: SquareCircuit<G::Scalar>,
+struct TestROM<G1, G2, S>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  S: StepCircuit<G2::Scalar> + Default,
+{
   rom: Vec<usize>,
-  _p: PhantomData<G>,
+  _p: PhantomData<(G1, G2, S)>,
+}
+
+#[derive(Debug, Clone)]
+enum TestRomCircuit<F: PrimeField> {
+  Cubic(CubicCircuit<F>),
+  Square(SquareCircuit<F>),
+}
+
+impl<F: PrimeField> StepCircuit<F> for TestRomCircuit<F> {
+  fn arity(&self) -> usize {
+    match self {
+      Self::Cubic(x) => x.arity(),
+      Self::Square(x) => x.arity(),
+    }
+  }
+
+  fn synthesize<CS: ConstraintSystem<F>>(
+    &self,
+    cs: &mut CS,
+    pc: Option<&AllocatedNum<F>>,
+    z: &[AllocatedNum<F>],
+  ) -> Result<(Option<AllocatedNum<F>>, Vec<AllocatedNum<F>>), SynthesisError> {
+    match self {
+      Self::Cubic(x) => x.synthesize(cs, pc, z),
+      Self::Square(x) => x.synthesize(cs, pc, z),
+    }
+  }
 }
 
-impl<G: Group> CircuitSet<G> for TestROM<G> {
-  fn num_augmented_circuits(&self) -> usize {
+impl<G1, G2> NonUniformCircuit<G1, G2, TestRomCircuit<G1::Scalar>>
+  for TestROM<G1, G2, TrivialSecondaryCircuit<G2::Scalar>>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+{
+  fn num_circuits(&self) -> usize {
     2
   }
+
+  fn primary_circuit(&self, circuit_index: usize) -> TestRomCircuit<G1::Scalar> {
+    match circuit_index {
+      0 => TestRomCircuit::Cubic(CubicCircuit::new(circuit_index, self.rom.len())),
+      1 => TestRomCircuit::Square(SquareCircuit::new(circuit_index, self.rom.len())),
+      _ => unimplemented!(),
```

Comment:
Nit: Is `unimplemented` the right approach, considering `num_circuits` states plainly the call is invalid?